### ISSUE TITLE
API for data provided by React

### ISSUE THREAD

**Issue by Rsl1122** [VolmitSoftware/React#363]
Hello.

I've been getting suggestions on Discord to add support for React to my plugin, [Plan](https://github.com/Rsl1122/Plan-PlayerAnalytics)

Is there/will there be an API for the data React provides?

If there is no long term storage, I'm sure some real time getters would suffice. My plugin could to do the storing.

Thanks, Rsl

**Comment by cyberpwnn** [comment]
This is one reason i don't obfuscate. I don't think it really needs an api since it is very easy to just use react as it is. 

### Get a sampled value
```java
// Get the FORMATTED value
SampledType.TPS.get().get();

// Get the RAW value
SampledType.TPS.get().getValue();

// Get the formatter of this sampler
SampledType.TPS.get().getFormatter();
```

### Queue an action to be fired
```java
// Let's select an area to fire the action on
SelectorPosition pos = new SelectorPosition();
pos.add(Bukkit.getWorld("world")); 
pos.add(player.getLocation().getChunk()); 

// Let's also specify entities in a filter. 
SelectorEntityType types = new SelectorEntityType(SelectionMode.WHITELIST);
types.add(EntityType.PIG);

// We specify who called the action (for responses and messages)
// You can also use new EmptyActionSource() for no logging
IActionSource source = new ConsoleActionSource(); // or PlayerActionSource

// Then we queue the action 
React.instance.actionController.fire(ActionType.PURGE_ENTITIES, source, pos, types);
```

**Comment by cyberpwnn** [comment]
Reopen if you still want some sort of wrapper or something, or you have other questions.

**Comment by cyberpwnn** [comment]
There is a graphing dataset you could probably tap into, but due to the amount of samplers and speed needed to iterate through them, i had to implement lossy compression. For example a graph of TPS will only have 2,000 recordings, however it wont delete the oldest sample, instead it will remove every other sampled value BEFORE the median of recorded time.

This means each sampler will have a HIGHLY accurate graph for about 10 seconds (20 samples per second), however 20 seconds back you will only see 10 samples per second. 30 minutes back you may only see 1 sample every few seconds.

**Comment by Rsl1122** [comment]
Sounds like it would be very easy to implement some tasks that store average values on my plugin for each minute.

I'm going to assume by wrapper you mean a separate class that could be distributed also in its own jar (similar to Litebans) - I don't have a copy of react, so that, or a dev copy would help.

Thanks!

**Comment by cyberpwnn** [comment]
Alright ill publish a wrapper to nexusvolmit when i get the chance (ill update this issue when im finished)

**Comment by cyberpwnn** [comment]
I think a wrapper would be smarter, however for the time being pm me on discord so you can get react in your CP

**Comment by LogGits** [comment]
:o also add plugin ram usage getter 👍

**Comment by Rsl1122** [comment]
I tried to pm you on discord but it was blocked due to "no common d
servers" - I have to start sleeping so I left a friend request.

Here is my spigot if you need it for the copy
https://www.spigotmc.org/members/rsl1122.122894/

**Comment by Rsl1122** [comment]
Looks like tapping into the graphing dataset worked, at least while a map is being held by someone. - I'll test further to see if I need a second task that takes a sample directly without the graphing data.

**Comment by cyberpwnn** [comment]
Probably a way to reduce memory usage (and garbage) when no one is using the map. I have a better method of getting data (soontm)

**Comment by cyberpwnn** [comment]
```java
React.instance.dataLogController.getDla().pullLayers(fromMS, toMS, maxIndices, SampledType.TPS);
```

This gives high fidelity data across huge date ranges. Use this sparingly however as it will give GC a run for it's money.

Implemented in 6.550
